### PR TITLE
Feature play with a second player remotely

### PR DESCRIPTION
**Pull request by slax57**
## Content

- Use connected user's sessionID to identify the 2 players of a game, and allow them to play only if it's their turn
- Display link to share to the second player for them to join in
- Auto-refresh the page every 5 seconds in order to see the board update automatically
- Minor CSS improvements
- Had to disable the Cypress tests for lack of time 🙁 

Trello card: https://trello.com/c/Mqau79qh/8-2-as-norbert-i-want-to-be-able-to-play-with-patrick-who-is-playing-on-another-computer

### CODE REVIEW

File: src/app.service.ts [hunk 1]
```diff
@@ -43,25 +52,61 @@ export class AppService {
...
       state: this.getBoardStateFromFile(),
+      player1: await this.usersRepository.save(player1),
     });
-    game = await this.gamesRepository.save(game);
-    return game;
+    return await this.gamesRepository.save(game);
+  }
+
+  async joinGame(gameId: number, player2SessionId: string): Promise<Game> {
+    const player2 = this.usersRepository.create({
```
Comment (fzaninotto) [inline src/app.service.ts]: you shouldn't crerate a player before checking that the game doesn't already have 2

**Comment by atilbian** [comment]
You need to rebase the branch on main !

File: src/app.controller.ts [hunk 2]
```diff
@@ -23,38 +23,46 @@ export class AppController {
...
-    const newGameId = (await this.appService.createNewGame(gameParams.size)).id;
+  async createNewGame(@Body() gameParams: { size: number }, @Req() req: Request) {
+    const newGameId = (await this.appService.createNewGame(gameParams.size, req.sessionID)).id;
     return { url: `/game/${newGameId}` };
   }
 
+  @Get('game/:id/join')
+  @Redirect('/')
+  async joinGame(@Param('id') id: number, @Req() req: Request) {
+    const gameId = (await this.appService.joinGame(id, req.sessionID)).id;
```
Comment (fzaninotto) [inline src/app.controller.ts]: I find it weird that you use the return value while you know the game id from the request. It makes me think that there is a hidden logic in appService.joinGame, but there is not...

File: src/handlebars/helpers.ts [hunk 3]
```diff
@@ -14,9 +14,23 @@ export function registerHandlebarsHelpers() {
     }
     return options.inverse(this);
   });
+  handlebars.registerHelper('cellCanBePlayed', function (cellValue, readyToPlay, currentPlayerTurnToPlay, options) {
+    if (cellValue === "empty" && readyToPlay && currentPlayerTurnToPlay) {
+      return options.fn(this);
+    }
+    return options.inverse(this);
+  });
+  handlebars.registerHelper('falseOrFalse', function (value1, value2, options) {
```
Comment (fzaninotto) [inline src/handlebars/helpers.ts]: name not clear

File: views/game.hbs [hunk 4]
```diff
@@ -2,30 +2,49 @@
...
-      <p>Current player :</p>
-      <p class="current-player" data-player="{{state.turn}}">{{state.turn}}</p>
+      {{#areEqual readyToPlay true}}
+      <p>
+        Current player:
+        <span class="current-player" data-player="{{game.state.turn}}">{{game.state.turn}}</span>
+      </p>
+      {{#areEqual currentPlayerTurnToPlay false}}
+      <p>Waiting for other player to make their move...</p>
+      {{/areEqual}}
```
Comment (fzaninotto) [inline views/game.hbs:24]: indentation is weird

File: src/handlebars/helpers.ts [hunk 5]
```diff
@@ -14,9 +14,23 @@ export function registerHandlebarsHelpers() {
     }
     return options.inverse(this);
   });
+  handlebars.registerHelper('cellCanBePlayed', function (cellValue, readyToPlay, currentPlayerTurnToPlay, options) {
+    if (cellValue === "empty" && readyToPlay && currentPlayerTurnToPlay) {
+      return options.fn(this);
```
Comment (fzaninotto) [inline src/handlebars/helpers.ts]: this is really obscure. Perhaps you need TS types & JS doc to show how to use it

File: views/game.hbs [hunk 6]
```diff
@@ -2,30 +2,49 @@
...
       <div class="top-edge"></div>
       <div class="bottom-edge"></div>
       <div class="hex-board-container">
         <i></i>
-        {{#each state.board}}
+        {{#each game.state.board}}
         {{#each this}}
-        {{#areEqual this.value 'empty'}}
-        <form method="post" action="/game/{{../../id}}">
+        {{#cellCanBePlayed this.value ../../readyToPlay ../../currentPlayerTurnToPlay }}
```
Comment (atilbian) [inline views/game.hbs]: You can use @root to access directly the top level object !

Comment (fzaninotto) [inline views/game.hbs]: these paths (`../../readyToPlay`) in a template are really unnatural. I don't understand how it works. Can't you use the `game` variable?

File: views/game.hbs [hunk 7]
```diff
@@ -37,7 +56,7 @@
         {{else}}
         <div class="cell" data-color="{{this.value}}">
         </div>
-        {{/areEqual}}
+        {{/cellCanBePlayed}}
```
Comment (fzaninotto) [inline views/game.hbs:68]: the template is becomeing hard to read. Time to split it?

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: src/handlebars/helpers.ts (hunk 3 above)
Comment (atilbian) [inline src/handlebars/helpers.ts]: We could make it a less generic func, e.g needToWaitForTurn(isGameStarted,isMyTurn) ?

File: views/game.hbs (hunk 4 above)
Comment (atilbian) [reply views/game.hbs:24]: I agree but it's fixed by the default conf of prettier..

File: views/game.hbs (hunk 7 above)
Comment (slax57) [reply views/game.hbs:68]: indeed it would be a great time to do it... we'll see if we got time later on 🙂

**Pull request merged**